build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1288,6 +1315,218 @@ message WaitExecutionRequest {
+      // does not match the filesystem. Otherwise the hash of the string constant "P" must
+      // be added to the prospective strong fingerprint.
+      ABSENT_FILE_OR_DIRECTORY_PROBE = 5;
+    }
+
+    // File or directory path.
+    string Path = 1;
+
+    // Defines the meaning of `Path` and how to compare to the local filesystem.
+    OpKind Kind = 2;
+
+    // For entries of `OpType.DIRECTORY_ENUMERATION`, the enumeration pattern used.
+    // When null/empty/missing the search pattern was "*" meaning the full directory.
+    string DirectorySearchPattern = 3;
+  }
+
+  // Observed path set entries. In order to ensure consistent hashing of the same `ObservedPathSet`,
+  // the entries MUST be sorted lexicographically by Path code point (or, equivalently, by UTF-8
+  // bytes).
+  repeated Entry Entries = 1;
+
+  // For directory comparisons in Entries marked with `OpKind.DIRECTORY_SEARCH`, defines the filenames,
+  // with extensions removed, that were read by the action. These are used as the filter for generating
+  // a directory hash to add to the prospective strong fingerprint.
+  repeated string ObservedAccessedFileNames = 2;
+}
+
+TODO: erik: Add AnyBuild 2-way exec contract and discuss VFS layouts
+
+// Represents a two-phase cache pathset, referenced or inlined, with the strong fingerprint
+// that can be used to retrieve build outputs if the pathset matches the client's filesystem.
+message ObservedPathSetAndStrongFingerprint {
+  // The digest for a blob containing the pathset, which can be retrieved from the
+  // [ContentAddressableStorage][build.bazel.remote.execution.v2.ContentAddressableStorage].
+  // Servers SHOULD increase the TTL of the referenced blob if necessary and
+  // applicable.
+  Digest pathset_digest = 1;
+
+  // An inlined copy of the pathset. The server SHOULD NOT inline pathsets unless requested by
+  // the client in the [GetPotentialActionResultsRequest][build.bazel.remote.execution.v2.GetPotentialActionResultsRequest]
+  // message. The server MAY omit inlining, even if requested, and MUST do so if inlining
+  // would cause the response to exceed message size limits.
+  bytes pathset = 2;
+
+  // The strong fingerprint generated along with the pathset from the filesystem after
+  // completion of the [Action][build.bazel.remote.execution.v2.Action]. This value is
+  // compared to the prospective strong fingerprint during comparison of
+  // an [ObservedPathSet][build.bazel.remote.execution.v2.ObservedPathSet] to the
+  // client filesystem.
+  Digest strong_fingerprint_digest = 3;
+
+  // The strong fingerprint digest of the [ActionResult][build.bazel.remote.execution.v2.ActionResult]
+  // to retrieve from [ActionCache.GetActionResult][build.bazel.remote.execution.v2.ActionCache.GetActionResult]
+  // if the prospective strong fingerprint generated from the pathset matches `strong_fingerprint_digest`.
+  Digest action_result_digest = 4;
+}
+
 // A request message for
```

Comment:
Consider cache epochs, and an API to request delete of a cache entry based on bad inputs

Reply:
epoch == salt as a concept

---
In reply to: [453828686](https://github.com/erikma/remote-apis/pull/1#discussion_r453828686) [](ancestors = 453828686)

Reply:
There's no concept for salt/epoch in the existing Bazel RE API. I just asked a question on the remote-exec working group external DL. In the interim I'm closing this comment as it's a separate concern from this proposal and can be handled separately.

Reply:
Proposed separately: https://github.com/bazelbuild/remote-apis/pull/149